src/Npgsql/ConnectorPool.cs

```diff
@@ -197,200 +175,188 @@ internal async ValueTask<NpgsqlConnector> AllocateLong(NpgsqlConnection conn, Np
-                        else
-                        {
-                            // Someone beat us to it
-                            newPruningTimer.Dispose();
-                        }
+                        // Physical open failed, decrement the open and busy counter back down.
+                        conn.Connector = null;
+                        Interlocked.Decrement(ref State.Open);
+                        throw;
                     }
-
-                    return connector;
                 }
 
-                if (state.Busy == _max)
+                // We may be racing a connector Close, either it decrements Open and Idle state between our reads
+                // or we read right after both decrements, in any case causing busy to be less than _max.
+                // When we read in between the decrements, idleCount will be > 0 and our next check in this loop will
+                // call TryAllocateFast, 'incorrect' but inconsequential and corrects once the Idle decrement happens.
+                // When we read right after all decrements, idleCount == 0 and openCount < _max so we loop again
+                // to open a new connector, if we lose that race we'll move into the waiter queue anyway.
+                // Keep in mind, busy is only really read after all other options, when the pool is exhausted.
+                // Pool exhaustion and idle connector closes together is a strange combination, and recollect that
+                // during Close any connectors that weren't idle don't decrement idle, obviating any race.
+                openCount = Volatile.Read(ref State.Open);
+                var idleCount = Volatile.Read(ref State.Idle);
+                var busy = openCount - idleCount;
+                if (busy == _max)
                 {
-                    // Pool is exhausted. Increase the waiting count while atomically making sure the busy count
-                    // doesn't decrease (otherwise we have a new idle connector).
-                    checked
+                    // Pool is exhausted.
+                    // Enqueue an allocate attempt into the waiting queue so that the next release will unblock us.
+                    var tcs = new TaskCompletionSource<NpgsqlConnector?>(TaskCreationOptions.RunContinuationsAsynchronously);
+                    _waiting.Enqueue((tcs, async));
+
+                    // Scenario: pre-empted waiter
+                    // Say there's a pre-emption of the thread right between our State.Busy read and our Enqueue.
+                    // If that happens and the waiter queue is empty before we enqueue we couldn't signal to any
+                    // releases we are a new waiter, causing any to add their connectors back into the idle pool.
+                    // We do a correction for that right here after our own enqueue by re-checking Idle.
+                    if (State.Idle > 0)
```

Comment:
Good one, this entirely focused on making sure waiters wouldn't get stuck with concurrent releasers but you're absolutely right it should also take into account physical close!

Reply:
Note that in this case, having an inconsistent view of Idle/Open may be more dangerous than above; not going into the condition above isn't too bad (we just continue down and possibly loop again). Not going into *this* condition could be more destructive... I may be wrong and it may be safe, but I'd like to avoid having to think about it if possible.

Reply:
I would model this as `Idle > 0 || Open < _max` so that'll be fine together with a dequeue on Close